core/src/main/java/org/elasticsearch/transport/TransportChannelHandler.java

```diff
@@ -0,0 +1,78 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.transport;
+
+import org.elasticsearch.common.logging.ESLogger;
+import org.elasticsearch.threadpool.ThreadPool;
+
+import java.io.IOException;
+
+/**
+ * Base class for delegating transport response to a transport channel
+ */
+public abstract class TransportChannelHandler<T extends TransportResponse> implements TransportResponseHandler<T> {
+
+    /**
+     * Convenience method for delegating an empty response to the provided changed
+     */
+    public static TransportChannelHandler<TransportResponse.Empty> emptyResponseHandler(ESLogger logger, String transportAction, TransportChannel channel, TransportRequest transportRequest) {
+        return new TransportChannelHandler<TransportResponse.Empty>(logger, transportAction, channel, transportRequest) {
+            @Override
+            public TransportResponse.Empty newInstance() {
+                return TransportResponse.Empty.INSTANCE;
+            }
+        };
+    }
+
+    private final ESLogger logger;
+    private final String transportAction;
+    private final TransportChannel channel;
+    private final TransportRequest transportRequest;
+
+    protected TransportChannelHandler(ESLogger logger, String transportAction, TransportChannel channel, TransportRequest transportRequest) {
+        this.logger = logger;
+        this.transportAction = transportAction;
+        this.channel = channel;
+        this.transportRequest = transportRequest;
+    }
+
+    @Override
+    public void handleResponse(T response) {
+        try {
+            channel.sendResponse(response);
+        } catch (IOException e) {
+            handleException(new TransportException(e));
+        }
+    }
+
+    @Override
+    public void handleException(TransportException exp) {
+        try {
+            channel.sendResponse(exp);
+        } catch (IOException e) {
+            logger.debug("failed to send failure, action [{}], request [{}]", e, transportAction, transportRequest);
```

Comment:
I would not take the logs we had in the Replication actions here. i'm not sure they make sense. Instead, I would add an option parameter for the constructor called `extraInfoOnError` and log here:

```
 "failed to send failure {}", extranInfo == null? "" : "("+ extraInfo +")"
```